File: Player/Windows/NewHyOn_Player/PlaybackModes/PlaybackContracts.cs
using System.Collections.Generic;
using SharedContentsInfoClass = AndoW.Shared.ContentsInfoClass;
using SharedPageInfoClass = AndoW.Shared.PageInfoClass;

namespace NewHyOnPlayer.PlaybackModes
{
    internal enum SeamlessLayoutState
    {
        Idle,
        Preparing,
        Ready,
        Active
    }

    internal enum SeamlessSlotState
    {
        Idle,
        Preparing,
        Ready,
        Active,
        Error
    }

    internal sealed class SeamlessContentItem
    {
        public SharedContentsInfoClass Source { get; set; }
        public string FilePath { get; set; }
        public NewHyOnPlayer.ContentType ContentType { get; set; }
        public int DurationSeconds { get; set; }
        public long ActualDurationSeconds { get; set; }
        public bool ShouldLoop { get; set; }
        public bool TransitionByTimer { get; set; }
        public int LoopDisableAfterEndCount { get; set; }
        public int TransitionEndEventCount { get; set; }

        public bool IsVideo
        {
            get { return ContentType == NewHyOnPlayer.ContentType.Video; }
        }
    }

    internal sealed class SeamlessSlotPlan
    {
        public string ElementName { get; set; }
        public bool IsMuted { get; set; }
        public double Width { get; set; }
        public double Height { get; set; }
        public double Left { get; set; }
        public double Top { get; set; }
        public int ZIndex { get; set; }
        public List<SeamlessContentItem> Items { get; set; }

        public SeamlessSlotPlan()
        {
            Items = new List<SeamlessContentItem>();
        }

        public bool HasPlayableItems
        {
            get { return Items != null && Items.Count > 0; }
        }
    }

    internal sealed class SeamlessPagePlan
    {
        public string PlaylistName { get; set; }
        public string PageName { get; set; }
        public double CanvasWidth { get; set; }
        public double CanvasHeight { get; set; }
        public int DurationSeconds { get; set; }
        public List<SeamlessSlotPlan> Slots { get; set; }

        public SeamlessPagePlan()
        {
            Slots = new List<SeamlessSlotPlan>();
        }

        public string PlanKey
        {
            get { return string.Concat(PlaylistName ?? string.Empty, "|", PageName ?? string.Empty); }
        }
    }

    internal sealed class SeamlessPlaybackStatus
    {
        public string ElementName { get; set; }
        public string CurrentContentName { get; set; }
        public string NextContentName { get; set; }
        public int CurrentIndex { get; set; }
        public int NextIndex { get; set; }
        public long ElapsedMilliseconds { get; set; }
        public long DurationMilliseconds { get; set; }
        public long ElapsedSeconds { get; set; }
        public long DurationSeconds { get; set; }
        public bool IsVisible { get; set; }
    }

    internal sealed class SeamlessPlaybackPulse
    {
        public SeamlessPlaybackStatus PrimaryContent { get; set; }
        public bool IsSecondTick { get; set; }
        public bool IsContentBoundary { get; set; }
    }

    internal sealed class SeamlessPagePlaybackRequest
    {
        public SharedPageInfoClass CurrentPage { get; set; }
        public SharedPageInfoClass NextPage { get; set; }
        public string PlaylistName { get; set; }
    }

    internal sealed class PlaybackDebugItem
    {
        public string ElementName { get; set; }
        public string CurrentContentName { get; set; }
        public string NextContentName { get; set; }
        public long ElapsedSeconds { get; set; }
        public long DurationSeconds { get; set; }
        public bool IsVisible { get; set; }
        public string LayoutName { get; set; }
        public int SlotIndex { get; set; }
        public SeamlessLayoutState LayoutState { get; set; }
        public SeamlessSlotState SlotState { get; set; }
    }
}

File: Player/Android/Quber4k/Quber_Player/app/src/main/java/kr/co/turtlelab/andowsignage/playback/PlaybackContracts.java
package kr.co.turtlelab.andowsignage.playback;

import java.util.ArrayList;
import java.util.List;

import kr.co.turtlelab.andowsignage.AndoWSignageApp;
import kr.co.turtlelab.andowsignage.datamodels.MediaDataModel;

public final class PlaybackContracts {

    private PlaybackContracts() {
    }

    public enum LayoutState {
        Idle,
        Preparing,
        Ready,
        Active
    }

    public enum SlotState {
        Idle,
        Preparing,
        Ready,
        Active,
        Error
    }

    public static final class ContentItem {
        public MediaDataModel source;
        public String fileName = "";
        public String filePath = "";
        public AndoWSignageApp.CONTENT_TYPE contentType = AndoWSignageApp.CONTENT_TYPE.None;
        public long durationSeconds = 1L;
        public long actualDurationSeconds = 1L;
        public boolean shouldLoop;
        public boolean transitionByTimer = true;
        public int loopDisableAfterEndCount;
        public int transitionEndEventCount;

        public boolean isVideo() {
            return contentType == AndoWSignageApp.CONTENT_TYPE.Video;
        }
    }

    public static final class SlotPlan {
        public String elementName = "";
        public boolean isMuted = true;
        public int width;
        public int height;
        public int left;
        public int top;
        public int zIndex;
        public final List<ContentItem> items = new ArrayList<>();

        public boolean hasPlayableItems() {
            return !items.isEmpty();
        }
    }

    public static final class PagePlan {
        public String playlistName = "";
        public String pageName = "";
        public double canvasWidth = 1920d;
        public double canvasHeight = 1080d;
        public int durationSeconds = 1;
        public final List<SlotPlan> slots = new ArrayList<>();

        public String getPlanKey() {
            return String.valueOf(playlistName) + "|" + String.valueOf(pageName);
        }
    }

    public static final class SlotSyncStatus {
        public String elementName = "";
        public String currentContentName = "";
        public String nextContentName = "";
        public int currentIndex = -1;
        public int nextIndex = -1;
        public long elapsedSeconds;
        public long durationSeconds = 1L;
        public boolean isVisible;
    }

    public static final class PlaybackPulse {
        public SlotSyncStatus status;
        public boolean isSecondTick;
        public boolean isContentBoundary;
    }
}
